### PR TITLE
Declare fused_moe/tuned_gemm preshuffled weight layout

### PR DESCRIPTION
**Pull request by yzhou103**
## Motivation
reported in this issue https://github.com/ROCm/aiter/issues/2194
`shuffle_weight()` records the preshuffled MFMA layout by setting an `is_shuffled`
attribute on the tensor it returns. A plain Python attribute is not tensor metadata,
so it does not survive tensor operations. `fused_moe` and `gemm_a16w16` then infer
"not preshuffled", select the `preshuffle_off` kernel and read preshuffled data with
the wrong layout — silently, with no error and no warning.
Measured on torch 2.10 / gfx950, the tag is lost by 8 of the 10 operations probed:
| operation | tag survives |
| --- | --- |
| `clone()`, `to(device, copy=True)`, `detach()`, `t[:]`, `view(dtype)` | no |
| `nn.Parameter(t)` | no |
| `copy.deepcopy(t)` | yes |
| `copy.deepcopy(nn.Parameter(t))` | no |
| `contiguous()` | yes, but only because it returns `self` when already contiguous |
Impact, identical config on each side:
- `fused_moe` with FP4 weights: the run-to-run noise floor is 1280 (stage2 reduces
  with atomics), while a cloned weight gives a max diff of **290816**
- `gemm_a16w16`, bf16 128x1024x512: a cloned weight gives a max diff of **139.0**,
  where the output magnitude is only ~22

## Technical Details

Add an optional `is_shuffled` argument so a caller can state the layout when the
tensor can no longer carry it:
```python
w = shuffle_weight(w_q, layout=(16, 16))
out = fused_moe(hidden, w.clone(), w2.clone(), ..., is_shuffled=True)
```
It defaults to `None`, which keeps inferring from the attribute, so no existing
caller is affected. Threaded through `fused_moe`, `fused_moe_`, `_fused_moe_impl`,
`fused_moe_2stages`, `fhmoe` and `gemm_a16w16`, with the matching `*_fake`
signatures updated so the `torch_compile_guard` schemas still line up.
`op_tests/test_moe_2stage.py` now passes the flag explicitly. Its i4x2 and int4
branches feed `shuffle_weight()`'s result through `pack_int8_to_packed_int4()` and
`.view()`, both of which return fresh tensors, so the tag was already gone before
the kernel saw it.

**Alternative considered**

Making the layout propagate on its own — a tensor subclass — is the better long-term
answer, and torchao's `AffineQuantizedTensor` shows the shape it should take: each
layout is its own type with a `Layout` / `TensorImpl` pair, `__torch_dispatch__`
intercepts every op, and `__tensor_flatten__` lets `torch.compile` see through it.

It is out of scope here for two reasons specific to a kernel library. First, a
subclass without dispatch interception is worse than none: probing one showed the tag
surviving `t[:]` and `view()`, and a slice of a preshuffled tensor is not a valid
preshuffled tensor, so the tag would keep propagating after it stopped being true.
Second, doing it properly does not close the loop from aiter alone. torchao owns the
module transform, so its subclass is what `load_state_dict` copies into; aiter does
not own module construction, and weights would still arrive as plain `nn.Parameter`
from the serving stack.

The explicit argument is needed regardless of that work: weights that arrive already
shuffled in a checkpoint, the tuner calling `ck_moe_stage*_fwd` directly, and the C++
op boundary all have no tensor able to carry a tag. `is_shuffled: bool | None` can
widen to accept a `Layout` later without breaking callers, so this is a step toward
that design rather than a competing one.

## Test Plan

`op_tests/test_is_shuffled.py` is new and covers both ops. It compares against the
untouched tensors instead of a fixed tolerance, and includes a negative control —
declaring `is_shuffled=False` for genuinely preshuffled weights must diverge — so
the test fails if the argument is ever ignored rather than passing by accident.
- `python op_tests/test_is_shuffled.py` passes
- `python op_tests/test_moe_2stage.py --no-flydsl-csv --quant 4`: 12/12 cases,
  `logits_diff` between 3.4e-06 and 1.2e-03, and the selected stage1 kernel is
  `..._MulABScaleShuffled_v3_...`, confirming the `preshuffle_on` variant is used

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: aiter/fused_moe.py [hunk 1]
```diff
@@ -685,7 +729,7 @@ def _fused_moe_impl(
         inter_dim * 2,
     ], f"Invalid MoE weight: {w1.shape=} {w2.shape=}"
     isG1U1 = inter_dim != w1.shape[1]
-    isShuffled = getattr(w1, "is_shuffled", False) or getattr(w2, "is_shuffled", False)
+    isShuffled = _resolve_is_shuffled(w1, w2, is_shuffled)
```
[the remark this answers is not in the document]
Comment (yzhou103) [reply aiter/fused_moe.py:738]: Thanks, but `metadata.run_1stage` does not invoke `ck_moe_stage1_fwd`.
It invokes `fused_moe_1stage`, whose ASM entry points do not inspect the
`is_shuffled` Python attribute; the resolved value is already used to select
the appropriate kernel metadata.

`ck_moe_stage1_fwd` belongs to the two-stage path, where
`fused_moe_2stages` attaches the explicit layout to w1/w2 before invoking
the CK stage wrappers. Therefore both paths already receive the layout
through the mechanism they use.

**Review by yifehuan — COMMENTED** [review]
what happens when is_shuffled=False hits a tuned CSV entry? For example, the DSV3 shape (token=1, model_dim=7168, inter_dim=256, E=257, topk=9) maps to an ExpertWeightShuffled stage-2 kernel. Since layout isn’t part of the config key, would we still launch that kernel for unshuffled weights?

**Comment by zufayu** [comment]
**Advisory review** (static + hand-run; not a merge gate). Validation/Perf ran no GPU stage — reasons are on their lines below. Findings tagged `[verified]` are traced to code/repro.

**Adds an explicit `is_shuffled`/`bpreshuffle` layout argument to fused_moe/fhmoe/gemm_a16w16 so a caller can declare preshuffled weights when tensor ops (clone/to/nn.Parameter) have dropped shuffle_weight()'s `is_shuffled` attribute, whose loss made dispatch silently infer "not preshuffled" and read preshuffled data with the wrong layout.**

Review (advisory): ⚠️ NEEDS WORK
Validation (deterministic): NOT RUN — auto-validation disabled (REVIEW_AUTO_VALIDATE=0); triage required the new target op_tests/test_is_shuffled.py (runtime surface: aiter/fhmoe.py, aiter/fused_moe.py, aiter/tuned_gemm.py)
Perf (advisory): NOT RUN — target exposes no benchmark entry point (no --scenario bench, no perftest harness); the change is dispatch-only (kwarg threading, two getattr calls, an occasional detach alias), no kernel/tiling/launch modified

⚠️ [verified] The declared layout cannot reach tuned-config kernel selection. The tuned-config key in aiter/fused_moe.py has no layout column (`_INDEX_COLS` at fused_moe.py:2159), so a tuned row's kernelName launches whatever `is_shuffled` says: at token=1, model_dim=7168, inter_dim=256, E=257, topk=9 the row in configs/model_configs/dsv3_fp4_tuned_fmoe.csv maps to a `moe_ck2stages_..._MulABScaleExpertWeightShuffled_v1_...` stage-2 kernel, and on `run_1stage` rows (e.g. the `_ps_` 1-stage asm kernels in tuned_fmoe.csv) the only warning, fused_moe.py:2426, is switched off by its `not run_1stage` condition. A caller declaring `is_shuffled=False` at such a shape still gets the Shuffled kernel — reviewer yifehuan asked exactly this on the PR and it is unanswered. If the declaration contradicts the tuned row, the wrong-layout read the PR exists to stop is [inferred] to persist, silently on 1-stage rows. **Author must** answer what a declaration that contradicts a tuned row does (reject, remap, or warn on both stages) and wire that in.
⚠️ [verified] `opus_weights_shuffled` bypasses the explicit declaration. fused_moe.py:898 computes it from the raw tensor attributes before any re-tag, while its twin read in `fused_moe_2stages` (fused_moe.py:3044) sits after `_tag_is_shuffled` (:3022) — the copy-paste left the `_fused_moe_impl` side unadapted. At an Opus-tuned shape (the `opus_moe2_afp8_wfp4_fp8_t64x256x256_sbm64_rbn3584` row for gfx950, 7168/512/E=384/topk=6 in configs/model_configs/dsv4_fp8fp4_tuned_fmoe.csv) with cloned weights plus `is_shuffled=True`, the first `get_2stage_cfgs` lookup discards the Opus row — default heuristics set the block_m that moe_sorting and the stage launches use — while the second lookup, after the re-tag, launches the Opus kernel: two lookups disagreeing inside one call. Whether that corrupts output or only costs the tuned config is [inferred]; the disagreement is in the code. **Author must** thread the resolved declaration into `opus_weights_shuffled` (or re-tag before the first lookup) so both lookups agree.
⚠️ [verified] The PR is stale against main and the rebase is not mechanical. The patch no longer applies (aiter/fhmoe.py:355, aiter/fused_moe.py:295); main moved 1579 lines in aiter/fused_moe.py since the PR base and grew new raw `is_shuffled` attribute readers the re-tag does not cover — e.g. the eligibility check at fused_moe.py:358 on main returns "weights are not both preshuffled" off raw getattr, so after rebase an explicit `is_shuffled=True` with copied weights would be declared ineligible for that path. Any CI result on the PR describes a tree that has moved. **Author must** rebase and re-derive every attribute reader the declaration has to reach, not just the two call sites fixed here.
📝 [verified] No consumer is wired. The kwarg on aiter/fused_moe.py changes nothing until a caller passes it, and the PR's own motivation table shows `nn.Parameter` — how serving stacks hold weights — drops the tag; no vLLM/SGLang/ATOM follow-up is mentioned. Until one adopts it, the silent wrong-layout read of issue #2194 persists downstream exactly as before. **Reviewer should ask** which downstream PR passes `is_shuffled` first and track it before merging this as a fix.
📝 [verified] The declaration's coverage is narrower than its tests. op_tests/test_is_shuffled.py's gemm half has no negative control — the MoE half asserts `is_shuffled=False` on shuffled weights must diverge, but nothing asserts `bpreshuffle=False` on a shuffled B diverges, so a tuned shape whose config ignores `bpreshuffle` passes vacuously; the fhmoe threading and the run_1stage path are exercised by no test in the PR; and the fhmoe shared-expert weights (shared_w1/shared_w2, viewed at ops/flydsl/fhmoe.py:300) have no declaration channel at all. A future refactor could drop the argument's effect without failing this suite. **Author must** add the gemm negative control or state why no shape distinguishes the two layouts, and say how shared-expert weights declare theirs.
